### PR TITLE
Avro serde support for unwrapped single-field schemas. Part of KLIP-3.

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description

This PR is looking to implement part of KLIP-3. It enhances the Avro serde classes to support (de)serialization of unwrapped single-field schemas.

This PR was pulled out of the monster #2899 PR. See the main PR for more details.

### Testing done
The usual.

### Reviewer checklist
  Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
  Ensure relevant issues are linked (description should include text like "Fixes #")

### CODE REVIEW

File: ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java [hunk 1]
```diff
@@ -29,6 +28,11 @@
 import org.apache.kafka.connect.data.Timestamp;
 import org.apache.kafka.connect.errors.DataException;
 
+/**
+ * Translates full set of Connect types to the limited subset supported by KSQL.
```
Comment (rodesai) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java]: This class also takes care of case-sensitivity.

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java [hunk 2]
```diff
@@ -53,236 +62,733 @@
...
+    expectedException.expect(SerializationException.class);
+    expectedException.expectCause(hasMessage(containsString(
+        "Cannot deserialize type boolean as type int32")));
+
+    // When:
+    deserializer.deserialize(SOME_TOPIC, bytes);
+  }
+
+  @Test
+  public void shouldDeserializedAvroBigInt() {
```
Comment (rodesai) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: We already have test cases for these (for e.g. see lines 817, 1062, 1072, etc) - let's consolidate. Ditto for deserializing into int

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java [hunk 3]
```diff
@@ -53,236 +62,733 @@
...
+    // Given:
+    givenDeserializerForSchema(Schema.OPTIONAL_INT64_SCHEMA);
+
+    final Map<org.apache.avro.Schema, Object> validCoercions = ImmutableMap.of(
```
Comment (rodesai) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: Should include int8, int16

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java [hunk 4]
```diff
@@ -53,236 +62,733 @@
...
+
+    // When:
+    final Object result = deserializer.deserialize(SOME_TOPIC, bytes);
+
+    // Then:
+    assertThat(result, is(value));
+  }
+
+  @Test
+  public void shouldDeserializedAvroMap() {
```
Comment (rodesai) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: nit: shouldDeserializeAvroMap (ditto for the other names)

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java [hunk 5]
```diff
@@ -53,236 +62,733 @@
...
+  private static final org.apache.avro.Schema DOUBLE_AVRO_SCHEMA =
+      parseAvroSchema("{\"type\": \"double\"}");
+
+  private static final org.apache.avro.Schema STRING_AVRO_SCHEMA =
+      parseAvroSchema("{\"type\": \"string\"}");
+
+  private static final org.apache.avro.Schema STRING_ARRAY_AVRO_SCHEMA =
+      parseAvroSchema("{\"type\": \"array\", \"items\": \"string\"}");
+
+  private static final org.apache.avro.Schema INT_MAP_AVRO_SCHEMA =
```
Comment (rodesai) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: We should also include connect's map with optional string keys representation for avro (list of key-value pairs). We have tests that cover this in the old structure (using `shouldDeserializeConnectTypeCorrectly`).

**Comment by big-andy-coates** [comment]
Thanks @rodesai - I agree on the test consolidation. I added the new to ensure the types could be handled as unwrapped single values, (something the old tests obviously didn't cover) and deliberately left the old tests inplace and untouched to prove the new code passed the old tests, with a view to consolidating in another PR. But happy to change in this PR given the build is green.

**Comment by rodesai** [comment]
@big-andy-coates I'm fine with doing it in a follow-up. In fact, as you say, staging it this way is nice because we have a record that we didn't break anything.

**Comment by big-andy-coates** [comment]
Merging with one reviewer as the main PR also has one approver

File: ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java]: Done in main PR. (Sorry, forgot to address before merging!)

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: Ah, difference is that the preexisting tests deserialize a BigInt _field_, right!

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java (hunk 3 above)
Comment (big-andy-coates) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: Avro doesn't have such types.  This test is _deserializing_ an avro type to Connect INT64. The only types we support deserializing into a BIGINT are Avro `int` and `long` if I'm not mistaken.

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java (hunk 5 above)
Comment (big-andy-coates) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: Done in follow up PR.

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java (hunk 4 above)
Comment (big-andy-coates) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: done in follow up Pr

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java (hunk 3 above)
Comment (rodesai) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: > We don't want to ever deserialize to an int8/16...

Connect can generate such types. E.g. I can use the connect avro converter to serialize a Schema.INT8_SCHEMA. Our tests should verify that we deserialize these correctly.

Comment (big-andy-coates) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: That doesn't make sense to me.  This is the deserializer test.  It tests we can deserialize avro types to the connect ones.   A test that checks we can serialize an INT8 into an avro int is testing the serializer, not the deserializer.

Comment (rodesai) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: Let's say I'm using Connect to write an int8 to a Kafka topic from some external source. Connect will write that data out as an avro int, and add some metadata to the avro schema describing the original type (int8). We should have a test that verifies that we can deserialize what Connect has written out into a KSQL field (of type INTEGER or BIGINT (or SHORT/INT16 or BYTE/INT8 in the future).

Comment (big-andy-coates) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: Ah - gotcha!  It was the metadata that I was missing.  Consider it done!

Comment (big-andy-coates) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java]: Done in second commit to #2941

**Pull request merged**